### PR TITLE
auto-register connected proxies at WorkloadGroup creation

### PR DESCRIPTION
**Pull request by stevenctl**
If I have a few connected proxies before I create the WorkloadGroup we can end up in a state where the entries are never registered. 

This is mostly a UX improvement. A user might not realize the order matters. 
Accidental deletion of a WorkloadGroup can trigger cleanup due to owner refs. Recreating the WorkloadEntries would be delayed without either this patch, or forcing a reconnect to ads from each VM.

### CODE REVIEW

**Comment by stevenctl** [comment]
should just get caught on reconnect on second thought..

**Comment by howardjohn** [comment]
We could make ordering not matter by rejecting connections with REGISTRATION_GROUP (or w/e the var is ) set and there is no WG. Then it will retry quickly with backoff until fixed?

**Comment by howardjohn** [comment]
Actually I think we already do that in your other PR: https://github.com/istio/istio/pull/44165/files#diff-807c450a5a764a90d4d31d652325d20dc8ca1dcbf46e4951cdabaabbcf1a9ef3R235

**Comment by costinm** [comment]
I agree with John's suggestion - if a VM has auto-register enabled, it
should fail connection if WG is missing. It is a security concern - I am
working on a PR to add more checks
on connect as well to tighten things up. WG defines what identity and
labels of the workload - we currently allow connection to specify arbitrary
labels and annotations without any
check, which may have dangerous results given our dependency on label
selectors for a lot of security-related things.

**Comment by stevenctl** [comment]
@costinm what about this case:

1. Proxy connect
2. Auto-registered WorkloadEntry is created
3. Group is deleted. Entry cleaned up by ownerRef
4. Group is recreated

There will be a potentially very long lag before the max connection time expires and the proxy reconnects to trigger recreation. 

Should we force-disconnect the proxies after step 3?

**Comment by stevenctl** [comment]
closing in favor of https://github.com/istio/istio/pull/45209

**Comment by costinm** [comment]
I don't think deleting a group is a common use case. 

Also not sure if WE should be deleted on group removal via ownerRef - but by Istiod which is the controller.

File: pilot/pkg/autoregistration/controller.go [hunk 1]
```diff
@@ -137,30 +133,32 @@ type HealthEvent = health.HealthEvent
...
-			store:            store,
-			cleanupLimit:     rate.NewLimiter(rate.Limit(20), 1),
-			cleanupQueue:     queue.NewDelayed(),
-			adsConnections:   map[string]uint8{},
-			maxConnectionAge: maxConnAge,
+	if !features.WorkloadEntryAutoRegistration && !features.WorkloadEntryHealthChecks {
+		return nil
+	}
+
+	if maxConnAge != math.MaxInt64 {
```
Comment (shashankram) [inline pilot/pkg/autoregistration/controller.go:140]: A comment on what this block is meant to do will help

File: pilot/pkg/autoregistration/controller.go [hunk 2]
```diff
@@ -120,11 +119,8 @@ type Controller struct {
 	// cleanupQueue delays the cleanup of auto registered WorkloadEntries to allow for grace period
 	cleanupQueue queue.Delayed
 
-	mutex sync.Mutex
-	// record the current adsConnections number
-	// note: this is to handle reconnect to the same istiod, but in rare case the disconnect event is later than the connect event
-	// keyed by proxy network+ip
-	adsConnections map[string]uint8
+	adsConnections        *adsConnections
+	lateRegistrationQueue controllers.Queue
```
Comment (shashankram) [inline pilot/pkg/autoregistration/controller.go:123]: nit: late/delayed/g

**Comment by stevenctl** [comment]
Not stale :()

File: pilot/pkg/autoregistration/controller.go [hunk 3]
```diff
@@ -215,14 +248,11 @@ func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) err
...
 		return nil
 	}
-	proxy.WorkloadEntryName = entryName
-	proxy.WorkloadEntryAutoCreated = autoCreate
 
-	c.mutex.Lock()
-	c.adsConnections[makeProxyKey(proxy)]++
-	c.mutex.Unlock()
+	proxy.SetWorkloadEntry(entryName, autoCreate)
+	c.adsConnections.Connect(conn)
```
Comment (hzxuzhonghu) [inline pilot/pkg/autoregistration/controller.go:253]: It is a little confusing, does it run this for all the connections? I think we should do this only for WLG related connections

File: pilot/pkg/autoregistration/connections.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/maps"
+)
+
+// Connection from a proxy to a control plane.
+// This interface exists to avoid circular reference to xds.Connection as well
+// as keeps the API surface scoped to what we need for auto-registration logic.
+type Connection interface {
```
Comment (hzxuzhonghu) [inline pilot/pkg/autoregistration/connections.go:30]: Make it as a inner type to reduce confusion with Connection struct

**Review by hzxuzhonghu — COMMENTED** [review]
Confirm the connection cache's efficiency, cache and operate on only what we care about

**Review by hzxuzhonghu — APPROVED** [review]

**Review by ericvn — APPROVED** [review]

**Pull request merged**